secboot/luks2/cryptsetup.go

```diff
@@ -165,51 +138,13 @@ func AddKey(devicePath string, existingKey, key []byte, options *AddKeyOptions)
 
 	args = append(args,
 		// container to add key to
-		devicePath,
-		// read new key from stdin.
-		// Note that we can't supply the new key and existing key via the same channel
-		// because pipes and FIFOs aren't seekable - we would need to use an actual file
-		// in order to be able to do this.
-		"-")
+		devicePath)
 
 	cmd := exec.Command("cryptsetup", args...)
-	cmd.Stdin = bytes.NewReader(key)
-
-	// Writing to the fifo must happen in a go-routine as it may block
-	// if the other side is not connected. Special care must be taken
-	// about the cleanup.
-	fifoErrCh := make(chan error)
-	go func() {
-		fifoErr := writeExistingKeyToFifo(fifoPath, existingKey)
-		if fifoErr != nil {
-			// kill to ensure cmd is not lingering
-			if cmd.Process != nil {
-				cmd.Process.Kill()
-			}
-			// also ensure fifo is closed
-			cleanupFifo()
-		}
-		fifoErrCh <- fifoErr
-	}()
-
-	output, cmdErr := cmd.CombinedOutput()
-	if cmdErr != nil {
-		// cleanupFifo will open/close the fifo to ensure the
-		// writeExistingKeyToFifo() does not leak while waiting
-		// for the other side of the fifo to connect (it may never
-		// do)
-		cleanupFifo()
-	}
-	fifoErr := <-fifoErrCh
-
-	switch {
-	case cmdErr != nil && (fifoErr == nil || errors.Is(fifoErr, syscall.EPIPE)):
-		// cmdErr and EPIPE means the problem is with cmd, no
-		// need to display the EPIPE error
+	// cryptsetup will read first the unlocking key, then the new key
+	cmd.Stdin = bytes.NewReader(bytes.Join([][]byte{existingKey, key}, []byte{'\n'}))
```

Comment:
two comments:

* keys can contain "\n" afaik, as they can come directly from crypto/rand.Read
* this code is a fork of some internal secboot code, so that would have the same problem we see here

Reply:
The issue with new lines have been solved by using `--key-file - --keyfile-size <key_size>`. I need to propose this change to secboot now.